components/_util/hooks/useDeepMemo.ts

```diff
@@ -0,0 +1,30 @@
+import * as React from 'react';
```

Comment:
It's no different with prev code. Why not use `React.useMemo` and realize memo self? ConfigProvider only need to care the passed prop is same ref and should not compare in deep. It will case additional perf issue since developer may wrap many ConfigProvider in real code. Shallow equal is good enough:
```tsx
// No additional render
const pageHeader  = { ghost: true };
<ConfigProvider pageHeader={pageHeader} />

// Additional render
<ConfigProvider pageHeader={{ ghost: true }} />
```

Reply:
It would be nice to just use useMemo and shallow compare. But it would introduce a breaking change. Consider the following situation（[one of the test case](https://github.com/ant-design/ant-design/pull/28792/files#diff-90fc6ceffa60c11a3b5f758ae2514534599c065943ad3507bd61d2110c8fadd4R58)）:

```jsx
const App = () => {
      const [, forceRender] = React.useReducer(v => v + 1, 1);
      const pageHeader = React.useRef({ ghost: false });
      return (
        <ConfigProvider pageHeader={pageHeader.current}>
          <button
            type="button"
            onClick={() => {
              pageHeader.current.ghost = true;
              forceRender();
            }}
            className="mutable"
          >
            Change Direction
          </button>
          <MemoedSibling />
        </ConfigProvider>
      );
    };
```
If user change the props mutablely. In the current antd version, this change will be delivered to the Consumer. Because ConfigProvider generate new object **every single render**. But if we use useMemo, this kind of mutable change will not trigger Consumer's re-render.

We may have users change props in mutable way. So theoretically, the shallow equal implementation would introduce a bug .

So if you thought it is a bad prictice(it is) and should not be supported, I will use the shallow equal implementation.

Reply:
antd v4 only promise immutable data for correct render behavior.

https://ant.design/docs/react/faq#Why-does-modifying-props-in-mutable-way-not-trigger-a-component-update